base-server.d.ts

```diff
@@ -0,0 +1,759 @@
+   *
+   * ```js
+   * server.on('error', error => {
+   *   trackError(error)
+   * })
+   * ```
+   *
+   * @param event The event name.
+   * @param listener The listener function.
+   * @returns Unbind listener from event.
+   */
+  on(event: 'fatal' | 'clientError', listener: (err: Error) => void): void
+  on<Action extends LoguxUserAction = LoguxUserAction>(
+    event: 'error',
+    listener: (err: Error, action: Action, meta: LoguxMeta) => void
+  ): void
+  on(
+    event: 'connected' | 'disconnected',
+    listener: (server: ServerClient) => void
+  ): void
+  on<Action extends LoguxAction | LoguxUserAction = LoguxUserAction>(
+    event: 'preadd' | 'add' | 'clean',
+    listener: (action: Action, meta: LoguxMeta) => void
+  ): void
+  on<Action extends LoguxUserAction = LoguxUserAction>(
+    event: 'processed',
+    listener: (
+      action: Action,
+      meta: LoguxMeta,
+      latencyMilliseconds: number
+    ) => void
+  ): void
+  on(
+    event: 'subscribed',
+    listener: (
+      action: LoguxSubscribeAction,
+      meta: LoguxMeta,
+      latencyMilliseconds: number
+    ) => void
+  ): void
+
+  /**
+   * Stop server and unbind all listeners.
+   *
+   * ```js
+   * afterEach(() => {
+   *   testServer.destroy()
+   * })
+   * ```
+   *
+   * @returns Promise when all listeners will be removed.
+   */
+  destroy(): Promise<void>
+
+  /**
+   * Define action type’s callbacks.
+   *
+   * ```js
+   * server.type('CHANGE_NAME', {
+   *   access (ctx, action, meta) {
+   *     return action.user === ctx.userId
+   *   },
+   *   resend (ctx, action) {
+   *     return { channel: `user/${ action.user }` }
+   *   }
+   *   process (ctx, action, meta) {
+   *     if (isFirstOlder(lastNameChange(action.user), meta)) {
+   *       return db.changeUserName({ id: action.user, name: action.name })
+   *     }
+   *   }
+   * })
+   * ```
+   *
+   * @param name The action’s type.
+   * @param callbacks Callbacks for actions with this type.
+   */
+  type<Action extends LoguxUserAction = LoguxUserAction>(
+    name: Action['type'],
+    callbacks: LoguxActionCallbacks<Action>
+  ): void
+
+  /**
+   * Define callbacks for actions, which type was not defined
+   * by any {@link Server#type}. Useful for proxy or some hacks.
+   *
+   * Without this settings, server will call {@link Server#unknownType}
+   * on unknown type.
+   *
+   * ```js
+   * server.otherType(
+   *   async access (ctx, action, meta) {
+   *     const response = await phpBackend.checkByHTTP(action, meta)
+   *     if (response.code === 404) {
+   *       this.unknownType(action, meta)
+   *       retur false
+   *     } else {
+   *       return response.body === 'granted'
+   *     }
+   *   }
+   *   async process (ctx, action, meta) {
+   *     return await phpBackend.sendHTTP(action, meta)
+   *   }
+   * })
+   * ```
+   *
+   * @param callbacks Callbacks for actions with this type.
+   */
+  otherType<Action extends LoguxUserAction = LoguxUserAction>(
```

Comment:
Generics is also useless for `otherType` since it will accept different types of actions